src/Adapter/Product/CommandHandler/UpdateProductOptionsHandler.php

```diff
@@ -83,8 +83,10 @@ public function handle(UpdateProductOptionsCommand $command): void
         );
 
         $isVisibleOnSearch = $this->productIndexationUpdater->isVisibleOnSearch($product);
-        if ($wasVisibleOnSearch !== $isVisibleOnSearch) {
-            $this->productIndexationUpdater->updateIndexation($product);
+        // Cannot perform this optimization check for multi-shop context, because product is always loaded from single shop
+        // so it would end up checking one shop product and leaving all other shops unhandled
+        if (!$shopConstraint->getShopId() || $wasVisibleOnSearch !== $isVisibleOnSearch) {
+            $this->productIndexationUpdater->updateIndexation($product, $shopConstraint);
```

Comment:
Ok I get the idea, but then no need to check isVisibleOnSearch beforehand, so the code could be more explicit
```php
        if ($shopConstraint->forAllShops()) {
            $this->productIndexationUpdater->updateIndexation($product, $shopConstraint);
        } else {
            $isVisibleOnSearch = $this->productIndexationUpdater->isVisibleOnSearch($product);
            if ($wasVisibleOnSearch !== $isVisibleOnSearch) {
                $this->productIndexationUpdater->updateIndexation($product, $shopConstraint);
            }
        }
```

Reply:
its not only for all shops but All shops and for group.
But ok, ill make it abit more explicit in my way.